jkube-kit/build/service/buildpacks/src/main/java/org/eclipse/jkube/kit/service/buildpacks/BuildPackCliDownloader.java

```diff
@@ -0,0 +1,154 @@
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at:
+ *
+ *     https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.jkube.kit.service.buildpacks;
+
+import org.apache.commons.lang3.StringUtils;
+import org.eclipse.jkube.kit.common.KitLogger;
+import org.eclipse.jkube.kit.common.util.FileUtil;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.nio.file.StandardCopyOption.COPY_ATTRIBUTES;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.eclipse.jkube.kit.common.util.EnvUtil.findBinaryFileInUserPath;
+import static org.eclipse.jkube.kit.common.util.EnvUtil.isMacOs;
+import static org.eclipse.jkube.kit.common.util.EnvUtil.isProcessorArchitectureARM;
+import static org.eclipse.jkube.kit.common.util.EnvUtil.isWindows;
+import static org.eclipse.jkube.kit.common.util.EnvUtil.getUserHome;
+import static org.eclipse.jkube.kit.common.util.IoUtil.downloadArchive;
+import static org.eclipse.jkube.kit.common.util.SemanticVersionUtil.removeBuildMetadata;
+
+public class BuildPackCliDownloader {
+  private static final String JKUBE_PACK_DIR = ".jkube";
+  private static final String PACK_UNIX_CLI_NAME = "pack";
+  private static final String PACK_WINDOWS_CLI_NAME = "pack.exe";
+  private static final String PACK_DEFAULT_CLI_VERSION_PROPERTY = "version";
+  private static final String PACK_CLI_LINUX_ARTIFACT = "linux.artifact";
+  private static final String PACK_CLI_LINUX_ARM64_ARTIFACT = "linux-arm64.artifact";
+  private static final String PACK_CLI_MACOS_ARTIFACT = "macos.artifact";
+  private static final String PACK_CLI_MACOS_ARM64_ARTIFACT = "macos-arm64.artifact";
+  private static final String PACK_CLI_WINDOWS_ARTIFACT = "windows.artifact";
+
+  private final KitLogger kitLogger;
+  private final String packCliVersion;
+  private final Properties packProperties;
+  private final File jKubeUserHomeDir;
+
+  public BuildPackCliDownloader(KitLogger kitLogger, Properties packProperties) {
+    this.kitLogger = kitLogger;
+    this.packProperties = packProperties;
+    packCliVersion = (String) packProperties.get(PACK_DEFAULT_CLI_VERSION_PROPERTY);
+    jKubeUserHomeDir = new File(getUserHome(), JKUBE_PACK_DIR);
+  }
+
+  public File getPackCLIIfPresentOrDownload() {
+    try {
+      File pack = findPreviouslyDownloadedPackCLIPath();
+      if (pack == null) {
+        downloadPackCli();
+        pack = resolveBinaryLocation().toFile();
+      }
+      return pack;
+    } catch (IOException ioException) {
+      kitLogger.warn("Not able to download pack CLI : " + ioException.getMessage());
+      return getLocalPackCLI();
+    }
```

Comment:
1. (T) Pack binary exists and is valid: Verify path is returned
2. Pack binary doesn't exist
  1. Download binary
     1. Download fails
       1. (T) Local Pack Binary exists: Verify that local pack binary path is returned
       2. (T) Local Pack Binary doesn't exist: Verify exception is thrown
     2. Download succeeds
       1. (T) Verify path is returned
       2. (T) Verify path file exists and has the right size?